### ISSUE TITLE
GmstTweaker Patches GLOB only

### ISSUE THREAD

**Issue by Sharlikran** [wrye-bash/wrye-bash#335]
When testing the dev branch at https://github.com/wrye-bash/wrye-bash/commit/8262388e71bc0b894f9c52c601327c4bbb6906de it patches GLOB changes from the (GlobalsTweak, bush.game.GlobalsTweaks) part of the patcher, but no the GMST Tweaks from the (GmstTweak, bush.game.GmstTweaks) part of the patcher.

I checked this and it works fine for Oblivion.  It's just not working for Skyrim for some reason.  When you have time can you look into that please?

**Comment by Utumno** [comment]
Any idea when that broke or was it broken from the start ( 9599368dee429f4b63f116d25945beae73d838db ) ? Can you give me minimal patch options that reproduce this ?

**Comment by Sharlikran** [comment]
Hippy to help.  Since there is no plugin I'll get a screenshot and I'll test that commit.

Sorry I'm a bit behind making theose other plugins for the names patcher. I got busy with skyrim mods and had to replace both right and left bearings and breaks on my car yesterday.

Hope you enjoyed your weekend so far, talk to you soon.

**Comment by Utumno** [comment]
Thanks - let's see when that broke - are you familiar with git bisect ?

**Comment by Sharlikran** [comment]
No, I have learned quite a lot from you, git wise. Not everything you would like of course. :) I'm interested mostly in playing and modding games right now. I keep mentioning that. I'd appreciate it if I could stay out of the development end of things. I have to make code changes to have the functionality I need for the games for personal use and that's all I'm really willing to do.

**Comment by Utumno** [comment]
Yes I see - but you can help and if you do then the functionality you need would be in Bash much faster. Plus if you make the code changes in a way I can readily review and merge (that is if you keep your branches up to date with current stable and correct dev code) that's - help. What's the use to make changes that are half correct and will anyway be left behind and you will be missing all the fixes/enhancements ? While if you just but follow closely upstream dev and with the experience you have with git |(and your skills as a tester and slowly as a programmer)

----
So yes for the current issue just test if it works on the initial patcher commit. If yes then it broke afterwards and bisect would help to pin the exact commit it broke. You would save me time by doing the bisect yourself but I understand what you say and I can do it myself (by letting behind the work I do on b2a - which goes quite fast thanks to bae code by jownd ;)

Once I finish with b2a I will then look at the saves for skyrim se and we are good to go

**Comment by Sharlikran** [comment]
```Git
G:\wrye-bash\wrye-bash [(790efbe...)|BISECTING]> git bisect good
warning: ignoring broken ref refs/remotes/origin/HEAD
warning: ignoring broken ref refs/remotes/origin/HEAD
81450fc8654a13f865d12a5c1ce9bc621801f068 is the first bad commit
commit 81450fc8654a13f865d12a5c1ce9bc621801f068
Author: MrD <the.ubik@gmail.com>
Date:   Wed Jun 15 12:57:08 2016 +0200

    MobObjects.keepRecords optimization:

    Mopy/bash/record_groups.py:
    @@ -264,5 +264,6 @@ def keepRecords(self,keepIds):
             """Keeps records with fid in set keepIds. Discards the rest."""
    -        self.records = [record for record in self.records if (record.fid ==
 (
    -            record.isKeyedByEid and GPath(bosh.modInfos.masterName),
    -            0) and record.eid in keepIds) or record.fid in keepIds]
    +        nullFid = (bosh.modInfos.masterName, 0)
    +        self.records = [record for record in self.records if
    +                        (record.fid in keepIds if not record.isKeyedByEid e
lse
    +                        (record.fid == nullFid and record.eid in keepIds))]


    We called GPath in a loop for a potentially huge number of records,
    plus readability suffered (still in pain).
    Took the opportunity to rename local variables and such to single out
    bosh.modInfos.masterName (Path actually) and dropped unneeded GPath
    calls on it.

:040000 040000 beda7cdb4bfc93a56c1127a78e161c00949a58f4 aa9f241a21a4da232d86e6ba
9359dc214b54ab63 M      Mopy
G:\wrye-bash\wrye-bash [(790efbe...)|BISECTING]>
```

**Comment by Sharlikran** [comment]
Interesting to see isKeyedByEid because that is in this part of the Skyrim [GMST Code](https://github.com/wrye-bash/wrye-bash/blob/dev/Mopy/bash/game/skyrim/records.py#L3090-L3093) which @lojack5 put there himself.

**Comment by Utumno** [comment]
Oh impressive :)

Hmm just spent the day in a rather frustrating battle with BSA infos so I don't have energy to look at this but I will surely do ;)

**Comment by Utumno** [comment]
Just to be sure - can you post me a setting that does not work ?

**Comment by Utumno** [comment]
Tried for instance with 'Msg: Soul Captured!' but seems to work - which GMST you were having problems with ?

**Comment by Sharlikran** [comment]
You are so cute. ;) You would pick the one that is most likely to work. It's not even a real tweak.

So Timescale is the only GLOB so you can choose that to make sure it keeps working after your fix.

For GMST choose Actor Strength Encumbra
        (

**Comment by Utumno** [comment]
I am not only cute I am smart too :D - I pinned it - will post a fix on my wip branch

**Comment by Sharlikran** [comment]
Sooy on my phome with the wife and kids. Work on this tomorrow. Go enjoy Halloween.

**Comment by Utumno** [comment]
Actually I think I managed to read the b2a's - that was like 8 hours of concentrated work, like non stop -  and was going to plug that into bash - turned out (after 3 hours) that I can't. You can't imagine my frustration (and my anger). I kind of knew it cause the FileInfos API does need refactoring (I am tackling it these last months in relation to #205 #247  and #326) but now the time came I realize it's a real hard nut.

Well - day lost - I can't even test the b2a's. AGGGRRR

Anyway I am still on holidays but it's one of the times I wonder why I spent so much time on Bash in general.

I guess I'll just patch up the API tomorrow - I will need a skyrim se bsa to throw that in too - if you come round to it PM me one now that I am on it - let's get done with bsas

--- 
Re: issue at hand: Really good job with the bisect btw - real helpful.

**Comment by Sharlikran** [comment]
Well thanks for the work on the BA2 files even if it didn't seem productive.  Getting BA2 support, and support with lz4 for the new BSA format is important but if you are on holiday, to be honest, we need to be patient.  For now I have ways to extract the strings files so Wrye can build a Bash Patch.  It's not ideal but it works for now.  

Try to enjoy your Holiday as much as you can.  Take care.

**Comment by Utumno** [comment]
OTH on holidays I have time (and energy) to do some real work... But yes it was a bit too much, in the last some years.

I pushed the fix in my wip branch - I don't understand the logic exactly so i will keep an eye open but this fixes at least the regression - let me close this when I commit on dev (as usual).

And Sharlikran - comments like "You would pick the one that is most likely to work", on a bad day, won't quite work. You do understand I hope that I am not trying to "pick a patcher that works". So please avoid those comments cause they may lead to misunderstandings. I trust you as a tester, I always take seriously your reports and to try to prove the code right and the user an idiot is bad taste, to say the least.  On the other hand a good report can really chop off hours (or months) - and actually having done the bisect for me was more than I expected :+1: (you should feel proud, that's rather advanced git compared to your beginnings couple years ago). Just on the next regression start by telling me which patcher failed :P

----

Well it's 4 o'clock here and been programming since noon so I am calling it a day. As I note in the commit message I though that if record.isKeyedByEid then its eid would be in keepIds - turns out that not and I can't be sure that's intentional.

**Comment by Utumno** [comment]
And thanks :)

**Comment by Sharlikran** [comment]
Thanks I'll let you know how it goes.  Calling it a day is good.  Take care.  I found another bug so don't try to resolve them all at once.  They are there for when you have time.  I'm not trying to make you at my beckon call anyway.

Off topic of this issue, but later or tomorrow can you remind me how to make a .pkl file?  I know you want to throw that away and that's fine if it's best but while it's still there and in use I want to make one for Fallout 4.

**Comment by Sharlikran** [comment]
The GLOB didn't stick and a few GMST didn't appear for Skyrim.  After testing these are the four that don't work.

![gmsterror1](https://cloud.githubusercontent.com/assets/1912426/19878158/addcc1de-9fa2-11e6-8a1b-9a8fcbed3122.jpg)

I'm going to test Oblivion to see how it does.

**Comment by Sharlikran** [comment]
Also I don't know what that isKeyedByEid really does.  Eid to me would mean the EDID subrecord.  I mean I guess you could comment it out and see what it does because it's only there for Skyrim.  Load this for Oblivion and uncheck everything but Tweak Settings.  Then check all the boxes just to get a result.  Don't be trying to trace it in PyCharm.  When you open the Bash Patch in TES4Edit, you will see some that overwrite records in the the ESM files and some that don't.  

The ones that don't overwrite a previous record are know as injected records.  They are pretty worthless unless the Skyrim/Oblivion EXE knows they are there and uses that variable.  However, I can see other mods, scripts, and conditions picking up on them if that's in the plugin.

I don't see the point of keying anything against the EDED field.  It's doesn't seem like anyone ever did it for Oblivion so why invent that for Skyrim?  About the only reason I an see to key against that is so you can match up the tweak with another mod.  So say there is a mod that has that value in it already.  Instead of injecting into Skyrim.esm, overwrite the record in the plugin.

I'm going to give you some test plugins. I'd like you to comment this patcher to explain what changed and that I suggested it.  If someone complains you can blame it on me.  I say this patcher is stupid how it works.  It needs to work a little differently.

Let me build the test plugins.

**Comment by Sharlikran** [comment]
Okay initial testing shows it's not really working for Oblivion from what I can see.  I could be wrong but how this patcher works is a mystery to me.  So throw away that [isKeyedByEid ](https://github.com/wrye-bash/wrye-bash/blob/dev/Mopy/bash/game/skyrim/records.py#L3093) for Skyrim.

**Comment by Sharlikran** [comment]
Okay I just thought of something else just right now.  As I am looking through the syntax of the global or tweak I noticed something.  There is no FormID.

Global (GLOB)
```python
    (_(u'Timescale'),_(u'Timescale will be set to:'),
        u'timescale',
        (u'1',         1),
        (u'8',         8),
        (u'10',       10),
        (u'12',       12),
        (u'18',       18),
        (u'24',       24),
        (u'[30]',     30),
        (u'40',       40),
        (_(u'Custom'), 0),
        ),
```
Game Setting (GMST)
```Python
    (_(u'Arrow: Litter Count'),
     _(u'Maximum number of spent arrows allowed in cell.'),
        (u'iArrowMaxRefCount',),
        (u'[15]',      15),
        (u'25',        25),
        (u'35',        35),
        (u'50',        50),
        (u'100',      100),
        (u'500',      500),
        (_(u'Custom'), 15),
        ),
```

The Global 'Timescale' exists so Wrye assigns it the FormID from the plugin it belongs to which for Oblivion.esm is 0000003A.  For Oblivion none of the DLC has 'iArrowMaxRefCount' so that record will become injected into Oblivion.esm.  And assigned some random FormID.  Unless a mod uses that EDID for a script or condition (CTDA) and unless someone decoded the game EXE and found that hidden in there then Oblivion will never use it.  IMO.

**Comment by Sharlikran** [comment]
@leandor What are your thoughts about this.  What does the CBash code suggest for Oblivion?  Does CBash compare EDID subrecords, assign the FormID and keep it to write into the Bash Patch?  Does it discard any GLOB or GMST that has no matching EDID from other files?

**Comment by Sharlikran** [comment]
@Zilav @Hugues92 This patcher has me questioning how things work.  Do the two of you feel that if Wrye Bash makes an injected record into Oblivion.esm or Skyrim.esm.  When the record doesn't exist in any DLC and nothing such as the EXE, Scripts, CTDA, or anything else references it, isn't that record just there, but does nothing?

**Comment by zilav** [comment]
Maybe in Oblivion GMST works only when injected into Oblivion.esm similar to worldspaces, that's why patcher does that. Definitely not in an issue in later games where GMST works just fine without injections.

**Comment by Sharlikran** [comment]
Well that's good to know but what about the other consern of mine? If I just make up a variable like iMaxNumberOfFollowers and set that to 5 I would think that Skyrim won't use it if the EXE the dlc or something else doesn't reference it? Am I right about that? I can't just make up any variable name I want.  If I make iMaxNumberOfFollowers it won't just magically make Skyrim use that variable and then I can have 5 followers.

**Comment by zilav** [comment]
Obviously.

**Comment by Utumno** [comment]
Thanks @Sharlikran it's important we clear up those misunderstandings - it's that I am not a native speaker too so...

To the isuue at hand - I don't know how this patcher works but it seems to work (no ?). If not we need another issue where we discuss all that.

I keep a note of these points but I don't know myself the code - to solve this I run the debugger and here's what I saw:

![335 -1](https://cloud.githubusercontent.com/assets/1365079/19890170/aafab09e-a041-11e6-80f5-e5b11cf345ce.jpg)

----

Ah yes - offtopic - I still need some skyrim SE bsas to test those, when you come round to it

**Comment by Utumno** [comment]
Re: pickle files - I don't know what's in there actually - I think there is a setting menu for generating them - yep `App_GenPickle` and "Generate PKL File" is the menu item

**Comment by Utumno** [comment]
And the record up close:

![335 -3](https://cloud.githubusercontent.com/assets/1365079/19891585/74bd360c-a049-11e6-9ee1-34c200b640b5.jpg)

That's the patcher in final PatchFile.buildPatch with only tweak u'Actor Strength Encumbrance Multiplier' selected

**Comment by Sharlikran** [comment]
I'll have to find a way to explain this better. Zilav confirmed my suspicions that you keep the Tweak when the GMST (Game Setting) or GLOB (Global) record exists in one of the active or imported plugins, and discard it when the record does not exist.  The Key or the subrecord you should look for is the EDID subrecord.

For GMST we need the Eid field.  From the Class I see the Eid referenced [here](https://github.com/wrye-bash/wrye-bash/blob/dev/Mopy/bash/brec.py#L1747) and there is a def [GetGMSTFid](https://github.com/wrye-bash/wrye-bash/blob/dev/Mopy/bash/brec.py#L1768)

````Python
for listofplugins do:
    if record.eid == tweak.eid:
        formIDtoAssign = brec.getGMSTFid(record.eid) <<-- found in Skyrim.esm get FormID
        <keep the tweak write to Bash Patch>
    else:
        <discard the tweak because no plugin uses the tweak>
````
```Python
    (_(u'Arrow: Litter Count'),
     _(u'Maximum number of spent arrows allowed in cell.'),
        (u'iArrowMaxRefCount',),  <<< tweak.eid
        (u'[15]',      15),
        (u'25',        25),
        (u'35',        35),
        (u'50',        50),
        (u'100',      100),
        (u'500',      500),
        (_(u'Custom'), 15),
        ),
```

**Comment by lojack5** [comment]
Wow, lot's activity on this one so...first an explanation on how the game deals with these things so you can understand why Bash is doing things this way, with a caveat that it's been a long while for me, and I can feel the details of the process fading from my memory. 

There are some records that are referenced in locations not by their FormId (stored in the record header, abbreviated FID in most of the code), but by their EditorId (stored in the EDID subrecord, abbreeviated EID in the code).  The most obvious example is in CDTA conditions and in scripts.  There the game references these objects by EID.  Say you had two different mods, each one created one of these special records with the same EID, but different FID.  The game in this case knows that both records are for the same thing, and reads them accordingly.  To complicate things further, there are some record types that the FID **still** matters to the game, even when some parts the EID is what matters.  This leads to very weird behavior if the above situation happened, because it will appear that some portions of the game are seeing one change, while others are not.  Even worse still there are some of these records that *do no exist* in the game's master files, and yet the *engine* recognizes them as valid records if the name shows up.  Finally, some of these records *do* show up in the game's main plugin, but with a NULL FID!  This is a mess, but Bash has a few things in place to deal with it.

So recap:
- Some records EID matters (good example is GMST, where the EID determines what kind of value is stored)
- Some records can have NULL FIDs.  In this case the game engine assumes it must be one of the allowed hardcoded records.
- This really only affects records that come in the base game, as NULL FIDs cause the engine to assume the record is from the base game.

Ok, `KeyedByEid` - what does this do?  When Bash is handling a record with a NULL FID, it checks to see if this should be considered an error.  I believe the actual check is in the phase before writing the file where we prune any records that are not actually needed in the patch.  If the record type has not been flagged as `KeyedByEid`, then Bash will throw out the record.  This means potentially valid GMSTs will get thrown out. (it does not get added to the `keepIds` set).  If however the record type is tagged with `KeyedByEid`, the the patcher will attempt to see if the EID is one of the ones recognized by the game.  This is where those PKL files come in.

The PKL file is generated by us and assigns a non-NULL FID to any GMST where:
- It appears in the base game file with a NULL FID, or
- It doesn't appear in the base game file, but it's hard coded into the engine.

Bash looks in the PKL file to see if we've assigned the record a FID.  If that's the case, then we just replace the record's FID with that one and carry on, allowing the change.  If not, the record keeps the NULL FID and gets thrown out eventually.

Additionally, in the patchers where Bash is editing GMSTs, etc.  Rather than check the FID vs previous mods to see if the change is needed (remember, Bash throws out records that will have no net effect, usually due to some previous mod also making the same change), Bash will check for records with the same EID, since that's what's important to the engine here.

Finally what records are affected:
- **GMST**: There are both hard coded and NULL FID GMSTs in the game, plus the record is directly referenced by scripts and CDTA subrecords via its EID, not FID.
- **GLOB**: I don't recall there being any NULL FID GLOBs, so iirc this record does not get the `KeyedByEid` setting.  However, EIDs are still important for this record, so the tweaker is set up to modify settings based on EID, not FID.
- **MGEF**: This one is a little foggy for me, but I do very specifically remember having to add the EID setting to CBash to make it handle correctly in a few rare circumstances.  Like I said it was actually years ago, before the move to GitHub even.  PBash didn't have any issues, and there are no NULL FID MGEF records, so it does not get the `KeyedByEid` setting, but some scripts do reference by EID, and a change internally to the CBash code had to be made.

In Skyim:
- **GMST**: Function the same as in Oblivion, so we also need a PKL file to go along with this
- **COBJ**: NULL FID COBJ records are allowed and do function properly, so this record gets flagged as `KeyedByEid`.
- Maybe more...we never really got Skyrim fully fleshed out before I needed to take more time for myself.

-------------------------

Long story short:
- It's complicated, and unfortunately I'm starting to forget some of the specifics
- `KeyedByEid` is used to allow some records to have NULL FIDs but still have meaning to Bash
- `keepIds` is a mechanism used by Bash to prune unneeded records in the final Bashed Patch.  It's related here because if a record has a NULL FID, normally it's not added to this set of items to be kept.
- 'keepIds' is also involved in how PBash merges mods into the Bashed Patch as well.

Hopefully that helps some, or at least point you in the right direction to dissect the code.

**Comment by Sharlikran** [comment]
How do I make a pkl file for Fallout 4? Where is the menu option or do I use a .py file from the command line? Do you remember the commands?

**Comment by Utumno** [comment]
Thanks @lojack5 - real helpful!

Code facts:

- isKeyedByEid is only set to true in MreCobj and MreGmst in _skyrim only_
- there are only 3 uses:
 - the one keeping the record IIUC

   ```python
    def setRecord(self,record):
        """Adds record to record list and indexed."""
        if self.records and not self.id_records:
            self.indexRecords()
        record_id = record.fid
        if record.isKeyedByEid:
            if record_id == (bosh.modInfos.masterName, 0):
                record_id = record.eid
        if record_id in self.id_records:
            oldRecord = self.id_records[record_id]
            index = self.records.index(oldRecord)
            self.records[index] = record
        else:
            self.records.append(record)
        self.id_records[record_id] = record
```

 - the one in this issue I got wrong cause I thought that if record has isKeyedByEid True then `record.fid in keepIds` would be false (erroneous assumption) - now the code reads:

  ```python
    def keepRecords(self,keepIds):
        """Keeps records with fid in set keepIds. Discards the rest."""
        self.records = [record for record in self.records if (record.fid == (
            record.isKeyedByEid and bosh.modInfos.masterName,
            0) and record.eid in keepIds) or record.fid in keepIds]
        self.id_records.clear()
        self.setChanged()
```
  in "MobObjects"

 - and the one in `patcher.patch_files.PatchFile#mergeModFile`:

   ```python
                nullFid = (bosh.modInfos.masterName, 0)
                ...
                if record.isKeyedByEid and fid == nullFid:
                    mergeIdsAdd(record.eid)
                else:
                    mergeIdsAdd(fid)
 ```
 

So bottom line it seems we that when we check we would accept the combo isKeyedByEid==False _and_ fid==(0,0) while this is never added to self.id_records (whatever that is). This (0,0) are for those ids that are not in master esm ? And do we ever get those ?

**Comment by Sharlikran** [comment]
I see people post that the Tweaks don't "sitck" and I have no idea why.  Which lead to me testing this.  It didn't seem to produce the results I was looking for so I opened a ticket.  Now I have no idea how I am going to make plugins to test the different situations.  So I'll try to make some minimal reproducer plugins.  

@Zilav, @Hugues92, where are all the places a GMST or GLOB can be referenced by the EDID only?

Also have either of you see any GMST or GLOB records with a nul FormID ever?  I ask because I explained to @Llojack5, many years ago, 2 or 3 if I remember, that Wrye has several places that reference a 32 bit integer as a FormID and added to the Fid table.  When that happens it's a null FormID.  The problem it was never a FormID to begin with it was just an integer.  For xEdit we use wbUnion but that doesn't exist in Wrye.  An example being [MelCoed](https://github.com/wrye-bash/wrye-bash/blob/dev/Mopy/bash/game/skyrim/records.py#L148-L154).

------
Oblivion: 
 - Plugin with GMST or GLOB in CTDA
 - MGEF with GMST or GLOB 
 - Script Records with GMST or GLOB 
Skyrim: 
 - Plugin with CTDA or GLOB in CTDA
 - Plugin with VMADor GLOB in CTDA
 - Plugin with COBJ or GLOB in CTDA

What am I missing and which is incorrect?  I want to cover all the possible solutions.

**Comment by zilav** [comment]
GMST is always referenced by EDID.
MGEF in Oblivion is referenced by EDID.
Everything else is referenced by FormID in all games.

**Comment by Sharlikran** [comment]
Do they ever have a nul FormID Zilav? Lojack is saying he has had to make conditions to handle that.  I have my doubts that there are Nul FormIDs but im really not the expert. I'm just trying to be thorough about this so Utumno doesn't have to revisit it if possible.

**Comment by zilav** [comment]
Anything that is not referenced can have NULL FormID in theory, including GMST, COBJ in Skyrim, NAVI, NOCM in FO4, etc.

**Comment by Utumno** [comment]
Still remains the problem of "engine only records"
